core-restore/src/test/scala/io/aiven/guardian/kafka/restore/RestoreClientInterfaceTest.scala

```diff
@@ -0,0 +1,150 @@
+  }
+
+  property("Round-trip with backup and restore") {
+    forAll(kafkaDataWithTimePeriodsGen()) { (kafkaDataWithTimePeriod: KafkaDataWithTimePeriod) =>
+      implicit val restoreConfig: RestoreConfig                               = RestoreConfig.empty
+      implicit val mockedKafkaProducerInterface: MockedKafkaProducerInterface = new MockedKafkaProducerInterface
+      implicit val kafkaClusterConfig: KafkaClusterConfig = KafkaClusterConfig(kafkaDataWithTimePeriod.topics)
+
+      val backupMock =
+        new MockedBackupClientInterfaceWithMockedKafkaData(Source(kafkaDataWithTimePeriod.data),
+                                                           PeriodFromFirst(kafkaDataWithTimePeriod.periodSlice),
+                                                           compression
+        )
+
+      backupMock.clear()
+      val calculatedFuture = for {
+        _ <- backupMock.backup.run()
+        _ <- akka.pattern.after(10 seconds)(Future.successful(()))
+        processedRecords = backupMock.mergeBackedUpData()
+        restoreMock      = new MockedRestoreClientInterface(processedRecords.toMap)
+        _ <- restoreMock.restore
+        restoredData = restoreMock.kafkaProducerInterface.producedData.asScala.toList
+      } yield restoredData
+
+      calculatedFuture.futureValue mustMatchTo kafkaDataWithTimePeriod.data
+    }
+  }
+
+  property("Round-trip with backup and restore works using fromWhen filter") {
+    forAll(
+      kafkaDataWithTimePeriodsAndPickedRecordGen(padTimestampsMillis =
+                                                   Range.inclusive(1, ChronoUnit.HOURS.getDuration.toMillis.toInt),
+                                                 min = 1000,
+                                                 max = 1000
+      )
+    ) { (kafkaDataWithTimePeriodAndPickedRecord: KafkaDataWithTimePeriodAndPickedRecord) =>
+      implicit val mockedKafkaProducerInterface: MockedKafkaProducerInterface = new MockedKafkaProducerInterface
+      implicit val kafkaClusterConfig: KafkaClusterConfig =
+        KafkaClusterConfig(kafkaDataWithTimePeriodAndPickedRecord.topics)
+      implicit val restoreConfig: RestoreConfig =
+        RestoreConfig(Some(kafkaDataWithTimePeriodAndPickedRecord.picked.toOffsetDateTime), None)
+
+      val backupMock =
+        new MockedBackupClientInterfaceWithMockedKafkaData(Source(kafkaDataWithTimePeriodAndPickedRecord.data),
+                                                           PeriodFromFirst(
+                                                             kafkaDataWithTimePeriodAndPickedRecord.periodSlice
+                                                           ),
+                                                           compression
+        )
+
+      backupMock.clear()
+      val calculatedFuture = for {
+        _ <- backupMock.backup.run()
+        _ <- akka.pattern.after(AkkaStreamInitializationConstant)(Future.successful(()))
+        processedRecords = backupMock.mergeBackedUpData()
```

Comment:
We are not passing in the `compression` here because its the job of the actual `RestoreClient` to decompress the backed up data later on.